feat: accept all args in `ddev composer create`, add `create-project` alias, fixes #6766

stasadev opened the pull request:
## The Issue

- #6766

## How This PR Solves The Issue

- Makes `ddev composer create` fully compatible (except for installing into a subdirectory) with `composer create-project`.
- Removes two `ddev restart` from `ddev composer create` (to speed up the process).
- Creates `docroot` (if it doesn't exist) inside `ddev composer create`, and runs `ddev restart` in this case.
- Adds `ddev composer create-project` alias to `ddev composer create`.
- Moves some logic from the main function to smaller functions.

## Manual Testing Instructions

Use quickstarts https://ddev.readthedocs.io/en/latest/users/quickstart/

Use `[<package> [<directory> [<version>]]]` syntax when running `ddev composer create`

```
# Works:
ddev composer create laravel/laravel
ddev composer create laravel/laravel .
ddev composer create laravel/laravel . "~11"
ddev composer create laravel/laravel . "^11"

# Fails:
ddev composer create laravel/laravel ..
ddev composer create laravel/laravel test
ddev composer create laravel/laravel test2 v11.0.0
```

The tests provide more complex scenarios.

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

## Release/Deployment Notes

Replacing `ddev composer create vendor/package` => `ddev composer create-project vendor/package .` in the documentation should only be done after a few releases, when most DDEV users have this feature.

rfay commented:
So we should be deprecating `ddev composer create` and allowing `ddev composer create-project` and just update all the quickstarts to use `ddev composer create ... .` ?

stasadev commented:
> So we should be deprecating `ddev composer create` and allowing `ddev composer create-project` and just update all the quickstarts to use `ddev composer create ... .` ?

The only difference I see is the behavior when no folder is specified:
```
$ composer create-project laravel/laravel
Creating a "laravel/laravel" project at "./laravel"
```

```
$ ddev composer create laravel/laravel
# will use the current directory
```

I'll add the alias `ddev composer create-project` to `ddev composer create` and we can decide what to do with it later, after most DDEV users have it, since we can't update our documentation too quickly as people with older versions of DDEV will complain.

rfay commented:
Rebased just to hopefully get rid of the annoying Windows failure.

rfay commented:
This is going to be brilliant, but of course it's super-sensitive territory and I don't think we should introduce it until after the v1.24.0 release. Let me know if you agree. Converted to draft.

stasadev commented:
I agree, this needs to be thoroughly tested, it's better not to include it in v1.24.0.
And as usual, there are problems with the Windows path in tests.

stasadev commented:
It passed the Windows tests for Composer.

rfay reviewed (changes requested):
I did only modest review and modest manual testing. but this PR should remove some really longstanding friction that DDEV has had with `composer create-project`

* The Laravel quickstart doesn't work with this PR, apparently some migrations don't get run? (It does work with DDEV v1.24.0). 
<img width="1165" alt="image" src="https://github.com/user-attachments/assets/5957faab-88b5-4da0-9149-6157ef6a770c">

* IMO in either this PR or a future PR `ddev composer create-project` should become the documented and standard technique, updating the quickstarts, but pointing people toward using `.` for the directory.  So `create-project` would be the command and `create` an alias. The docs will need to move around a bit because they won't have to do the dance around why `ddev composer create`.
* We'll have to figure out whether there are ways to keep people out of trouble. If they create their project in a subdirectory it's usually not a good idea.

stasadev commented:
> The Laravel quickstart doesn't work with this PR, apparently some migrations don't get run? (It does work with DDEV v1.24.0).

Thanks for catching that, it's because I removed a spare `ddev restart`, that updated the `.env` file with new database information. I will add it back, because it not only updates the configuration files, but also triggers hooks that users can add (`pre-start` and `post-start`). It was a bad decision to remove it, it is better to create a project slowly than to do it quickly and wrong.

> IMO in either this PR or a future PR `ddev composer create-project` should become the documented and standard technique, updating the quickstarts, but pointing people toward using `.` for the directory.

I think it should be done much later, replacing `ddev composer create vendor/package` => `ddev composer create-project vendor/package .` in the documentation should only be done after a few releases, when most DDEV users have this feature.

> We'll have to figure out whether there are ways to keep people out of trouble. If they create their project in a subdirectory it's usually not a good idea.

I can add a confirmation dialog when people provide a directory to `ddev composer create-project`.

stasadev commented:
About two `ddev restart`, that we have in `ddev composer create`, from Discord:

> The main reason for `ddev restart` is when the docroot changes or suddenly appears and wasn't there before, because the webserver needs to be restarted. It could be replaced probably with restarting the webserver.
So make sure you test it with lots of empty directories, and with mutagen enabled.

I decided to experiment with docroot and mutagen here.
I removed restart if the docroot exists, otherwise I create the docroot and do a restart.

If there is no restart, I run `MutagenSyncFlush`, `CreateSettingsFile`, and `PostStartAction`.
This should significantly increase the speed.

Needs more testing.

---

> The Laravel quickstart doesn't work with this PR, apparently some migrations don't get run? (It does work with DDEV v1.24.0).

Fixed.

> I can add a confirmation dialog when people provide a directory to `ddev composer create-project`

Added.

rfay commented:
I like the technique. Of course, the docroot may appear magically with any `composer create-project`, not just in the mutagen context.

stasadev commented:
And there was a `--create-docroot` flag. But now `docroot` is always created when you run `ddev config --docroot=<docroot>`, which is included with most of our quickstarts.

I tested several quickstarts with and without mutagen on Linux, and it worked fine.

Tests are faster too:

Tests / tests (pull-push-test-platforms, true) (pull_request) Successful in 162m (another PR)
Tests / tests (pull-push-test-platforms, true) (pull_request) Successful in 141m (this PR)

rfay commented:
> But now docroot is always created when you run ddev config --docroot=<docroot>, which is included with most of our quickstarts

Ah, thanks. Forgot about that path. 

Significant improvement on tests!

rpkoller reviewed (commented):
I've went through every project type that uses composer. i tried to always use the `.` notation and the first one or two project types that also had a version on composer create i experimented with the `v` and `~` and `^`....the results were not clear neither what the correct notation is. but figured it out, also thanks to the comment of randy noting i have to use quotes when i am using ~with a version number.... results are here: https://gist.github.com/rpkoller/d0fa272b81b7705dd3f56c35e01c8f8a  the only quickstarts that are broken are magento2, pimcore and sulu. the latter is mainly due to a change in their setup and they are providing a website.xml and the move step we recommend is obsolete. and i think it is necessary to clarify more the function of the name and key value. magento i will investigate some more but i run into a 404 error after following all the steps. and pimcore has an error on composer create. that "might" be either the quickstart being outdated or maybe related to the PR? but overall this looks more than solid. thank you for all the work that went into this!

rpkoller commented:
retested the failing magento and pimcore with DDEV HEAD, and there they are failing in the same way, so this PR has nothing to do with the fails, these quickstarts same as the outdated Sulu one are failing anyway.

rfay commented:
This is obviously very close, but I moved it to draft just so it doesn't get accidentally pulled before we are ready, probably after v1.24.1

tyler36 commented:
Seems to work. I think there is potential for developers mixing the `docroot` and `composer create` directories and getting `403` errors.

## Test: mixed messages

1. Initialize DDEV

```shell
ddev config --project-type=laravel --docroot=public
```

1. Install Laravel project

```shell
ddev composer create laravel/laravel gooey
```

1. Site generate 403 because `docroot: "public"` but I specified to install in `gooey` so docroot should be `gooey/public`

## Test: install current directory

1. Initialize DDEV

```shell
ddev config --project-type=laravel --docroot=public
```

1. Install Laravel project

```shell
ddev composer create laravel/laravel .
```

1. Confirm site works.

tyler36 reviewed (commented):
Works as designed.

stasadev commented:
> I think there is potential for developers mixing the `docroot` and `composer create` directories and getting `403` errors.

@tyler36, the issue is that `approot` is changed, and `docroot` is relative to `approot`, but we don't have an option to modify `approot`. We always assume `approot` is the folder where `.ddev` is located, and many things rely on that.

We also have a warning to notify users about this:

> Installing the project in the 'gooey' subdirectory is uncommon.
> DDEV won't properly apply CMS configuration in a subdirectory.
> This will be similar to running 'composer create-project' without DDEV.

For Laravel, this means it would use SQLite in the subdirectory instead of MariaDB (since the `gooey/.env` file won’t be updated).

---

> Works as designed.

Thank you for testing!

rfay reviewed (approved):
I confess to being a bit overwhelmed by this one, but it's got lots of miles and careful care and test coverage, thanks. We should get it in and get some miles on it.

I tested manually with some weird Drupal installs like

```
ddev composer create drupal/recommended-project
ddev composer create drupal/recommended-project . v9.5.9
ddev composer create drupal/recommended-project . ~9
```

They all worked perfectly.  Pull when ready!
